command/pipe.go

```diff
@@ -0,0 +1,355 @@
+		}
+
+		err := c.doUpload(ctx, dsturl)
+		if err != nil {
+			return &errorpkg.Error{
+				Op:  c.op,
+				Src: nil,
+				Dst: dsturl,
+				Err: err,
+			}
+		}
+		return nil
+	}
+}
+
+func (c Pipe) doUpload(ctx context.Context, dsturl *url.URL) error {
+	err := c.shouldOverride(ctx, dsturl)
+	if err != nil {
+		if errorpkg.IsWarning(err) {
+			printDebug(c.op, err, nil, dsturl)
+			return nil
+		}
+		return err
+	}
+
+	// override destination region if set
+	if c.dstRegion != "" {
+		c.storageOpts.SetRegion(c.dstRegion)
+	}
+	dstClient, err := storage.NewRemoteClient(ctx, dsturl, c.storageOpts)
+	if err != nil {
+		return err
+	}
+
+	metadata := storage.NewMetadata().
+		SetStorageClass(string(c.storageClass)).
+		SetSSE(c.encryptionMethod).
+		SetSSEKeyID(c.encryptionKeyID).
+		SetACL(c.acl).
+		SetCacheControl(c.cacheControl).
+		SetExpires(c.expires)
+
+	if c.contentType != "" {
+		metadata.SetContentType(c.contentType)
+	} else {
+		metadata.SetContentType(guessStdinContentType(dsturl))
+	}
+
+	if c.contentEncoding != "" {
+		metadata.SetContentEncoding(c.contentEncoding)
+	}
+
+	err = dstClient.Put(ctx, &stdin{file: os.Stdin}, dsturl, metadata, c.concurrency, c.partSize)
+
+	if err != nil {
+		return err
+	}
+
+	msg := log.InfoMessage{
+		Operation:   c.op,
+		Source:      nil,
+		Destination: dsturl,
+		Object: &storage.Object{
+			StorageClass: c.storageClass,
+		},
+	}
+	log.Info(msg)
+
+	return nil
+}
+
+// shouldOverride function checks if the destination should be overridden if
+// the destination object and given pipe flags conform to the
+// override criteria.
+func (c Pipe) shouldOverride(ctx context.Context, dsturl *url.URL) error {
+	// if not asked to override, ignore.
+	if !c.noClobber {
+		return nil
+	}
+
+	dstClient, err := storage.NewClient(ctx, dsturl, c.storageOpts)
+	if err != nil {
+		return err
+	}
+
+	dstObj, err := getObject(ctx, dsturl, dstClient)
+	if err != nil {
+		return err
+	}
+
+	// if destination not exists, no conditions apply.
+	if dstObj == nil {
+		return nil
+	}
+
+	var stickyErr error
+	if c.noClobber {
+		stickyErr = errorpkg.ErrObjectExists
+	}
+
+	return stickyErr
+}
+
+func validatePipeCommand(c *cli.Context) error {
+	if c.Args().Len() != 1 {
+		return fmt.Errorf("expected destination argument")
+	}
+
+	dst := c.Args().Get(0)
+
+	dsturl, err := url.New(dst, url.WithRaw(c.Bool("raw")))
+	if err != nil {
+		return err
+	}
+
+	// destination must be a file
+	if dsturl.IsBucket() || dsturl.IsPrefix() {
+		return fmt.Errorf("target %q must be a file", dsturl)
+	}
+
+	// wildcard destination can not be used with pipe
+	if dsturl.IsWildcard() {
+		return fmt.Errorf("target %q can not contain glob characters", dst)
+	}
+
+	return nil
+}
+
+func guessStdinContentType(dsturl *url.URL) string {
+	contentType := mime.TypeByExtension(filepath.Ext(dsturl.Absolute()))
+	if contentType == "" {
+		return "application/octet-stream"
+	}
+	return contentType
+}
```

Comment:
Not really guessing from stdin, maybe it could be renamed to something like `guessContentTypeByExtension`.

Reply:
It used to try to predict file type from stdin. It didn't work as I expected then I removed that logic. I should rename the function. Thanks for the pointing that out.